drivers/misc/bcm2835_smi.c

```diff
@@ -0,0 +1,1108 @@
+ *    specific prior written permission.
+ *
+ * ALTERNATIVELY, this software may be distributed under the terms of the
+ * GNU General Public License ("GPL") version 2, as published by the Free
+ * Software Foundation.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+ * IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/of.h>
+#include <linux/platform_device.h>
+#include <linux/of_address.h>
+#include <linux/of_platform.h>
+#include <linux/mm.h>
+#include <linux/slab.h>
+#include <linux/pagemap.h>
+#include <linux/dma-mapping.h>
+#include <linux/dmaengine.h>
+#include <linux/semaphore.h>
+#include <linux/spinlock.h>
+#include <linux/io.h>
+
+#define BCM2835_SMI_IMPLEMENTATION
+#include <linux/broadcom/bcm2835_smi.h>
+
+#define DRIVER_NAME "smi-bcm2835"
+
+#define USE_DMA 1
+/* max number of bytes where DMA will not be used */
+#define DMA_THRESHOLD_BYTES 128
+#define DMA_BOUNCE_BUFFER_SIZE (1024 * 1024 / 2)
+#define DMA_BOUNCE_BUFFER_COUNT 3
+
+#define N_PAGES_FROM_BYTES(n) ((n + PAGE_SIZE-1) / PAGE_SIZE)
+
+#define DMA_WRITE_TO_MEM true
+#define DMA_READ_FROM_MEM false
+
+struct bcm2835_smi_instance {
+	struct device *dev;
+	struct smi_settings settings;
+	__iomem void *smi_regs_ptr, *cm_smi_regs_ptr;
+	dma_addr_t smi_regs_busaddr;
+
+	struct dma_chan *dma_chan;
+	struct dma_slave_config dma_config;
+
+	struct {
+		struct semaphore callback_sem;
+		void *buffer[DMA_BOUNCE_BUFFER_COUNT];
+		dma_addr_t phys[DMA_BOUNCE_BUFFER_COUNT];
+		struct scatterlist *sgl[DMA_BOUNCE_BUFFER_COUNT];
+	} bounce;
+
+	struct scatterlist *buffer_sgl;
+
+	int clock_source;
+	int clock_divisor;
+
+	spinlock_t register_lock;
+	bool device_registered;
+};
+
+/****************************************************************************
+*
+*   SMI clock manager setup
+*
+***************************************************************************/
+
+#define cm_reg(reg) (&((struct cm_smi_register_map *)cm_smi_regs_ptr)->reg)
+
+static void smi_setup_clock(struct bcm2835_smi_instance *inst)
+{
+	__iomem void *cm_smi_regs_ptr = inst->cm_smi_regs_ptr;
+
+	dev_info(inst->dev, "Setting up clock...");
+	/* Disable SMI clock and wait for it to stop. */
+	writel(CM_PWD | 0, cm_reg(CTL));
```

Comment:
The cm_reg macro looks very convoluted.
This is a common pattern:
```c
static inline void smi_write_reg(struct bcm2835_smi_instance *inst, unsigned reg, u32 val)
{
  writel(val, inst->smi_regs + reg);
}

static inline u32 smi_read_reg(struct bcm2835_smi_instance *inst, unsigned reg)
{
  return readl(inst->smi_regs + reg);
}

```
And similar for the cm_smi_regs.